### PR TITLE
[Notifier] Add Infobip bridge

### PR DESCRIPTION
**Pull request by jeremyFreeAgent**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no 
| New feature?  | yes
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/13623

- [x] Remove default host

Add Infobip SMS.

see https://www.infobip.com

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/Infobip/InfobipTransportFactory.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+{
+    /**
+     * @return InfobipTransport
+     */
+    public function create(Dsn $dsn): TransportInterface
+    {
+        $scheme = $dsn->getScheme();
+        $username = $this->getUser($dsn);
+        $password = $this->getPassword($dsn);
+        $from = $dsn->getOption('from');
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Infobip/InfobipTransportFactory.php:35]: You should probably throw an exception if not passed (have a look at the Freemobile transport  factory for an example).

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Notifier/Bridge/Infobip/InfobipTransportFactory.php (hunk 1 above)
Comment (jeremyFreeAgent) [reply src/Symfony/Component/Notifier/Bridge/Infobip/InfobipTransportFactory.php:36]: I've added it but we do not have that exception thrown for Nexmo and Twilio. Do you want me to add it too for those?

**Comment by Jontsa** [comment]
Infobip recommends API key authentication which is more flexible than using basic auth. https://dev.infobip.com/#section/Authentication

**Comment by jeremyFreeAgent** [comment]
I have changed the auth.

Thank you @Jontsa

**Comment by jeremyFreeAgent** [comment]
I've added tests based on Freemobile too.

File: src/Symfony/Component/Notifier/Bridge/Infobip/InfobipTransport.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    {
+        return sprintf('infobip://%s?from=%s', $this->getEndpoint(), $this->from);
+    }
+
+    public function supports(MessageInterface $message): bool
+    {
+        return $message instanceof SmsMessage;
+    }
+
+    protected function doSend(MessageInterface $message): void
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Infobip/InfobipTransport.php:51]: Should return a `SentMessage` :)

File: src/Symfony/Component/Notifier/Bridge/Infobip/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+Infobip Notifier
+===============
+
+Provides Infobip integration for Symfony Notifier.
+
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Infobip/README.md:5]: You should probably document the DSN pattern here (we should do the same for all other transports if you have time to contribute that in another PR).

File: src/Symfony/Component/Notifier/Bridge/Infobip/README.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
+Infobip Notifier
+===============
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Infobip/README.md:2]: Missing = at the end.

File: src/Symfony/Component/Notifier/Bridge/Infobip/Tests/InfobipTransportFactoryTest.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        $factory = $this->initFactory();
+
+        $this->expectException(UnsupportedSchemeException::class);
+
+        $dsnUnsupported = 'unsupported://authtoken@default?from=0611223344';
+        $factory->create(Dsn::fromString($dsnUnsupported));
+    }
+
+    private function initFactory(): InfobipTransportFactory
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Infobip/Tests/InfobipTransportFactoryTest.php:64]: I would remove the method and inline the instance creation instead.

File: src/Symfony/Component/Notifier/Bridge/Infobip/Tests/InfobipTransportTest.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    public function testSendNonSmsMessageThrowsException(): void
+    {
+        $transport = $this->initTransport();
+
+        $this->expectException(LogicException::class);
+
+        $transport->send($this->createMock(MessageInterface::class));
+    }
+
+    private function initTransport(): InfobipTransport
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Infobip/Tests/InfobipTransportTest.php:47]: I would call it `getTransport()` (I've renamed it in Freemobile tests).

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier_transports.xml [hunk 7]
```diff
@@ -34,6 +34,10 @@
             <tag name="texter.transport_factory" />
         </service>
 
+        <service id="notifier.transport_factory.infobip" class="Symfony\Component\Notifier\Bridge\Infobip\InfobipTransportFactory" parent="notifier.transport_factory.abstract">
+            <tag name="texter.transport_factory" />
+        </service>
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier_transports.xml:39]: You should rebase on current master and move this to the new transport.php file.

File: src/Symfony/Component/Notifier/Bridge/Infobip/README.md (hunk 3 above)
Comment (jeremyFreeAgent) [reply src/Symfony/Component/Notifier/Bridge/Infobip/README.md:5]: Done!

File: src/Symfony/Component/Notifier/Bridge/Infobip/Tests/InfobipTransportFactoryTest.php (hunk 5 above)
Comment (jeremyFreeAgent) [reply src/Symfony/Component/Notifier/Bridge/Infobip/Tests/InfobipTransportFactoryTest.php]: Fixed!

File: src/Symfony/Component/Notifier/Bridge/Infobip/Tests/InfobipTransportTest.php (hunk 6 above)
Comment (jeremyFreeAgent) [reply src/Symfony/Component/Notifier/Bridge/Infobip/Tests/InfobipTransportTest.php]: Fixed!

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier_transports.xml (hunk 7 above)
Comment (jeremyFreeAgent) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier_transports.xml]: Done!

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @jeremyFreeAgent.

**Comment by fabpot** [comment]
@jeremyFreeAgent Can you submit a PR on symfony/recipes (you can have a look at other notifier bridge recipes to get some inspiration)?

**Comment by jeremyFreeAgent** [comment]
I've submitted the PR https://github.com/symfony/recipes/pull/812

**Pull request merged**